wgsl/analyze/test.py

```diff
@@ -1600,6 +1601,274 @@ def test_follow(self):
+        I = _sym("I")
+
+        L = _sym("L")
+        S = _sym("S")
+        E2 = _sym("E2")
+        E1 = _sym("E1")
+        E0 = _sym("E0")
+        List = _sym("LIST")
+        ListEnd = _sym("LISTEND")
+        Call = _sym("Call")
+        Template = _sym("Template")
+
+        LDef = _def("L", _star(S) )
+        SDef = _def("S", _seq(E2,Semicolon) )
+        E2Def = _def("E2", _choice( _seq(E2,Lesser,E1), _seq(E2,Greater,E1), E1 ) )
+        E1Def = _def("E1", _choice( E0, _seq(E1,Plus,E0) ) )
+        CallDef = _def("Call",_seq(I,LParen,List,RParen))
+        TemplateDef = _def("Template",_seq(I, Lesser, List, Greater, LParen, List, RParen ))
+        E0Def = _def("E0", _choice( _seq(LParen,E2,RParen),_seq(LBracket,E2,RBracket), _seq(LBrace,E2,RBrace), I, Call, Template) )
+        ListDef = _def("LIST", _optional(_seq(E2, _optional(List))))
+        ListEndDef = _def("LISTEND", _seq(Comma, _optional(List)))
+        defs = [IDef, LDef,SDef,E2Def,E1Def,E0Def,CallDef,TemplateDef,ListDef,ListEndDef]
+        g = _gl("L", *defs)
+        return g
+
+    def test_dump(self):
+        g = self.toy_grammar()
+        g.compute_first()
+        g.compute_follow()
+        #g.dump()
+        print(file=sys.stdout,flush=True)
+                
+    def test_lalr1(self):
+        g = self.toy_grammar(False)
+        parse_table = g.LALR1()
+        #parse_table.write(sys.stdout)
+        #print(file=sys.stdout,flush=True)
+        # This grammar has conflicts
+        self.assertTrue(parse_table.has_conflicts())
+
+class DragonBook_4_17(unittest.TestCase):
+    def test(self):
+        g = Grammar.Grammar.Load(DRAGON_BOOK_EXAMPLE_4_17,'E')
+        g.compute_first()
+        g.compute_follow()
+        #g.dump()
+        #print(file=sys.stdout,flush=True)
+
+# Example 4.34
+class DragonBook_4_34(unittest.TestCase):
+    def toy_grammar(self):
+        # Grammar 4.19 in example 4.34
+        """
+        language = E
+        E = E '+' T | T
+        T = T '*' F | F
+        F = '(' E ')' | 'id'
+        """
+
+        # Tokens
+        LParen = _fixed("(")
+        RParen = _fixed(")")
+        Plus = _fixed("+")
+        Times = _fixed("*")
+        Id = _fixed("id")
+
+        E = _sym("E")
+        T = _sym("T")
+        F = _sym("F")
+
+        EDef = _def("E", _choice(_seq(E,Plus,T),T))
+        TDef = _def("T", _choice(_seq(T,Times,F),F))
+        FDef = _def("F", _choice(_seq(LParen,E,RParen),Id))
+        g = _gl("E", EDef,TDef,FDef)
+        return g
+
+    def xtest_first_follow(self):
+        g = self.toy_grammar()
+        g.compute_first()
+        g.compute_follow()
+        #g.dump()
+        print(file=sys.stdout,flush=True)
+
+    def xtest_lalr1(self):
+        g = self.toy_grammar()
+        parse_table = g.LALR1()
+        #parse_table.write(sys.stdout)
+        #print(file=sys.stdout,flush=True)
+        self.assertFalse(parse_table.has_conflicts())
+
+class TemplateParsingMidsize(unittest.TestCase):
```

Comment:
This is a toy language with various statement kinds (if, for), and function call and templates. A type can be a template,  a template can be "called".

This **does not have LALR(1)** conflicts.  Which surprised me.  
This is the LALR(1) parse table.

[lalr.midsize-tempalte-lang.txt](https://github.com/gpuweb/gpuweb/files/10460962/lalr.midsize-tempalte-lang.txt)